Add variable resolution (#305)
unicode-org/message-format-wg: spec/formatting.md

# WIP DRAFT MessageFormat 2.0 Formatting Behaviour

## Introduction

This document defines the behaviour of a MessageFormat 2.0 implementation
when formatting a message for display in a user interface, or for some later processing.

The document is part of the MessageFormat 2.0 specification,
the successor to ICU MessageFormat, henceforth called ICU MessageFormat 1.0.

## Variable Resolution

To resolve the value of a Variable,
its Name is used to identify either a local variable,
or a variable defined elsewhere.
If a local variable and an externally defined one use the same name,
the local variable takes precedence.

> It is an error for a local variable definition to

Reviewer: The language is a bit awkward; I suggest:

Any reference to a local variable before it is defined is an error.

Author: @macchiati Good point; replaced with your suggestion.

> It is an error for a local variable definition to
> refer to a local variable that's defined after it in the message.

Author: ```suggestion
Any reference to a local variable before it is defined is an error.
```

The document is part of the MessageFormat 2.0 specification,
the successor to ICU MessageFormat, henceforth called ICU MessageFormat 1.0.

## Variable Resolution

To resolve the value of a Variable,
its Name is used to identify either a local variable,
or a variable defined elsewhere.
If a local variable and an externally defined one use the same name,

> the local variable takes precedence.

Reviewer 2: This implies that name shadowing is possible. My current thinking is that it should be OK, and I see use-cases for it. However, I don't recall an explicit discussion on the subject. It might also make sense to start with a stricter design and disallow shadowing at first.

Author: I rather think that local variables must be allowed to shadow externally provided named arguments.

For instance, I could well imagine translation systems with the following attributes:
1. Translators are able to define additional local variables for a message.
2. A developer may pass the same basket of arguments to multiple messages, not all of which use each of them.

If shadowing were not allowed, then a combination of the above could result in a runtime error which would be difficult to discover ahead of time.

Reviewer 3: I think that this part of the change (shadowing) is a bit of an overreach.
It is unrelated to the hoisting problem, and it was not commonly agreed.

In general I don't think it is a good idea to allow translators to override a developer decision.
So "Translators are able to define additional local variables for a message." is a cons in my book, not a pro.

And _"A developer may pass the same basket of arguments to multiple messages, not all of which use each of them."_ does not look like something related to shadowing.

There is a reason many linters warn about shadowing.

> "If shadowing were not allowed, then a combination of the above could result in a runtime error which would be difficult to discover ahead of time."

Not necessarily.
One can say that translators can't define variables (and that is a good thing).
It is an error, but the decision between runtime error and simply ignore the conflicting local variable is implementation dependent.
We should not use that as an argument for what the standard should do.

---

But TLDR: 

* I don't see valid use-cases for it. In fact the proposal we had used different sigils for local variables (macros at the time) and arguments.
* I don't recall an explicit discussion on the subject.
* It might also make sense to start with a stricter design and disallow shadowing at first.

---

Filed https://github.com/unicode-org/message-format-wg/issues/310
